README.md

```diff
@@ -1,6 +1,168 @@
+
+`public function getUsername(): string`
+
+Returns the username of this user.
+
+`public function getPasswordHash(): string`
+
+Returns the password of the user, hashed using the strategy returned above.
+
+## Usage
+
+You instantiate the `Auth` class with an instance of your `AuthUserGateway` and an instance of the `Session` interface.
+
+`$auth = new Auth($userGateway, $session);`
+
+*You do not need to implement the `Session` interface yourself, the package provides an implementation of this. For more information on how to use that implementation, see the Session section below*
+
+Once you have an instance of the `Auth` class, you can do a number of things.
+
+`$auth->login($username, $password)`
+
+If the username is found, this function compare the provided password with the stored password and return `true` or `false`.
+
+If the username is not found, the function will throw a `UserNotFound` exception.
+
+If the login is successful, this function will also set the cookie on the user's machine.
+
+If a user is already logged in, a `UserAlreadyLoggedIn` exception is thrown.
+
+`$auth->isLoggedIn()`
+
+Returns `true` is the user is already logged in, and `false` otherwise.
+
+`$auth->getCurrentUser()`
+
+Returns the `AuthUser` that is currently logged in.
+
+If you call this function but the user is not logged in, a `NoUserLoggedIn` exception is thrown.
+
+`$auth->logout()`
+
+Logs the current user out, and resets the cookie.
+
+## Hashing Strategies
+
+Each user will have a `PasswordHashingStrategy` that controls how their password is hashed before it is stored in your datastorage layer.
+
+The strategy has two functions:
+
+`function hashPassword(string $password): string`
+
+this returns a hashed version of the password, using the strategy
+
+`function verifyPassword(string $passwordToVerify, string $passwordHash): bool`
+
+This checks that the given password matches the hash.
+
+### Which strategy
+
+The package provides a number of implementations you can use.  If you are starting fresh, we suggest you use either the `PasswordLockPasswordHashingStrategy` or the `PHPPasswordHashingAPIStrategy`.
+
+Using the `PHPPasswordHashingAPIStrategy` is straightforward:
+
+`$strategy = new PHPPasswordHashingAPIStrategy(new PHPPasswordAPIWrapperImplementation());`
+
+If you want to use the `PasswordLockPasswordHashingStrategy`, then you will need to generate and store a key. You *must not* store this key in the same datastore that you keep the hashed passwords.
+
+`$key = Defuse\Crypto\Key::createNewRandomKey();`
+
+Then pass this key to the strategy:
+
+`$strategy = new PasswordLockPasswordHashingStrategy($key);`
+
+The `Key` class provides some functions to help you store this key:
+
+`$string = $key->saveToAsciiSafeString();`
+
+`$key = Defuse\Crypto\Key::loadFromAsciiSafeString($string);`
+
+### Migrating Strategies
+
+Need some info here!
```

Comment:
This might change soon anyway. I'm thinking of working on a Cognito user provider which also supports migration, but I may have to change the way this works in order to fit that flow.